### ISSUE TITLE
major reference ?

### ISSUE THREAD

**Issue by peternara** [tyandzx/caffe#8]
Hi,
I want to know your reference(paper) about triplet loss. 
Can you give me the paper you refer to? facenet??

**Comment by tyandzx** [comment]
That's right. We use cosine-metric other than l2-metric used in the paper.

**Comment by peternara** [comment]
@tyandzx  thanks. :)
I try to training using vgg deep face.
for example.( is  prototxt for training)

name: "VGG_FACE_16_layer"
layer {
  name:"data"
  type:"Data"
  top :"data"
  top : "label"
  include{
  phase: TRAIN
  }
,,,
,,
,,,
layer {
  bottom: "fc6"
  top: "fc7"
  name: "fc7"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  type: "InnerProduct"
  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "xavier"
      std: 0.005
    }
    bias_filler {
      type: "constant"
      value: 1
    }
  }
}
layer {
  bottom: "fc7"
  top: "fc7"
  name: "relu7"
  type: "ReLU"
}
layer {
  bottom: "fc7"
  top: "fc7"
  name: "drop7"
  type: "Dropout"
  dropout_param {
    dropout_ratio: 0.5
  }
}

## add layers for triplet loss and embedding layers
layer {
  name: "fc8_embed"
  type: "InnerProduct"
  bottom: "fc7"
  top: "fc8_embed"
  param {
    lr_mult: 1
    decay_mult: 0
  }
  inner_product_param {
    num_output: 256
    weight_filler {
      type: "xavier"
    }
  }
}
layer {
    name: "sample_triplet"
    type: "SampleTriplet"
    bottom: "fc8_embed"
    top: "triplet"
    sample_triplet_param {
      label_num: 4
      sample_num: 8
    }
 }
 layer {
    name: "triplet_loss"
    type: "TripletLoss"
    bottom: "fc8_embed"
    bottom: "triplet"
    top: "triplet_loss"
    triplet_loss_param {
      margin: 0.1
    }
    loss_weight: 1
 }
                     
In previous similar experiments, I keep failing. 
Using the learning model from the intermediate processing, I extracted the embedding layer(fc8_embed) 's feature and compared them using l2-distance.
I compared using various aligned crop face. But the result was a mess. :(

Could you tell me what the problem is?

ps. I applied a pre-learned model(softmax).

**Comment by tyandzx** [comment]
I'm not quite sure about the problem. Two points maybe you can check.
1. Checkout whether training loss can decrease. If so, you training procedure seems right. If not, perform gradient checking of related layers to see whether they work well.
2. Cause you are using l2-distance during testing, checkout whether your feature has been l2-normalized, which doesn't been normalized during training.

**Comment by peternara** [comment]
@tyandzx thanks.
I check it.
hmm, 

1.  during training 
Iteration 0, Testing net (#0)
Test net output #32: triplet_loss = 1.42264 (* 1 = 1.42264 loss)
..
Iteration 260, loss = 1.4725
Train net output #32: triplet_loss = 1.4725 (* 1 = 1.4725 loss)
...
Iteration 5000, Testing net (#0)
triplet_loss = 1.40002 (* 1 = 1.40002 loss)
..
Iteration 9000, lr = 0.001
Train net output #32: triplet_loss = 1.41331 (* 1 = 1.41331 loss)
..
Iteration 19960, lr = 0.001
Train net output #32: triplet_loss = 1.40535 (* 1 = 1.40535 loss)

Overall, I could see that it was very much reduced. But it feels very small.
Should we lower the learning rate?:)
ps. this network'softmax training is pretty well learned.


2. how to normalized method during training? normalized layer existed??
or, When measuring distances, should I normalize?

my distance source 
float l2_similarity(std::vector<float> feat1, std::vector<float> feat2)
{
        float A  = 0.f;
        for(uint i=0; i<feat1.size(); i++){
                float dis =(feat1[i]-feat2[i]);
                A+=(dis*dis);
        }
        return A > 0.0 ? sqrtf(A) : 0.0;
}

**Comment by tyandzx** [comment]
1. I think the triplet loss decreases very slow, maybe you can try smaller margin or reduce learning rate.
2. Cause we use cosine-metric during training, we'd better also use cosine-metric in test phase. Concretely, in your code you should calculate cosine distance between feat1 and feat2 and the bigger the better(more similar).

**Comment by peternara** [comment]
@tyandzx thanks.:)

1. I will try that.
2. I did not say it, but I had already tried it(=cosine distance). Since the learning did not go well, no measurement method seemed useless.

I will try to do a little more(margin or learning rate) and reply again when I am done.
thanks.:)
